api/submit_result.go

```diff
@@ -32,12 +36,32 @@ func (s *resultSubmitter) Submit(executionID string, outputKey string, outputDat
 	if err != nil {
 		return err
 	}
-	if err := exec.Complete(outputKey, outputData); err != nil {
-		return err
+
+	// parse output data into a map.
+	var outputDataMap map[string]interface{}
+	switch d := outputData.(type) {
+	case map[string]interface{}:
+		outputDataMap = d
+	case string:
+		if err = json.Unmarshal([]byte(d), &outputDataMap); err != nil {
+			err = fmt.Errorf("invalid output data error: %s", err)
+		}
+	default:
+		err = errors.New("unexpected output data type")
+	}
+
+	if err == nil {
+		err = exec.Complete(outputKey, outputDataMap)
 	}
+
+	if err != nil {
+		exec.Failed(err)
+	}
+
 	if err = s.api.execDB.Save(exec); err != nil {
 		return err
 	}
+
 	go pubsub.Publish(exec.Service.ResultSubscriptionChannel(), exec)
-	return nil
+	return err
```

Comment:
We could delegate all this error handling to another function to make sure we always publish the result, here we have a return on the save that might be necessary (and i think it's good to have it) but we might add some stuff in between and forget that this function should publish almost all the time.

I was thinking of something like that 
https://github.com/mesg-foundation/core/blob/05dd0762eb9dd37ece4d66f3bac68fff4875f201/api/submit_result.go#L30-L37

Reply:
done

Reply:
actually, i applied but reverted the changes you proposed. because if there is an error we still may need to publish it because it can be about unmarshalling the output data or invalid output key etc. here in your proposition we don't publish in case of error which we should sometimes.

Reply:
I re-implemented this again because it's now possible after the changes made for https://github.com/mesg-foundation/core/pull/596#discussion_r240911881. But still, we don't return on error and always publish it.